Adds support for local paths and project bundles

IanCaio opened the pull request:
This PRs introduces 2 changes:

- It allows the prefix `local:` to be used on the project file paths. That prefix translates to a local path in relation to that project file path.
- It allows saving a project with its resources (Project Bundle) from both the CLI (`lmms --makeBundle ~/lmms/projects/input.mmp ~/lmms/projets/bundle/output.mmp`) or the `Save As` dialog (by marking `Save as Project Bundle (with resources)`).

A project bundle is basically a project file and a _resources_ folder, containing all external files used by the project (i.e.: samples, presets, plugins). The logic for the creation of the bundle goes as following:

1) There's a `std::map` on DataFile.cpp called `ELEMENTS_WITH_RESOURCES`. It maps a QString (the TagName of the element) to a `std::vector` of QStrings (the attributes of that element that are paths to resources). New elements can be easily added to the map so they are also addressed when creating the bundle.
2) Create a directory called `resources` on the same path as the output bundle project
3) For each tag name on `ELEMENTS_WITH_RESOURCES`, check if any of the attributes that links to paths are present. If they are, get an absolute path to the resource, copy that resource to the `resources` directory created earlier and update the attribute to point to that copied file (`local:resources/filename`).

To avoid name conflicts between files inside the resource folder, all of them are renamed to a number followed by the extension of the file (i.e.: `1.wav`, `2.dll`).

![SaveAs](https://user-images.githubusercontent.com/22241596/97571513-917c6600-19c6-11eb-9e94-2770b326f37b.png)

It's not allowed to save a bundle on the same path where there's also another bundle. I first thought of allowing bundles being overwritten, but I'm aiming towards simplicity on this PR: Overwritting requires that I remove the existing _resources_ folder, which is fine. Except that there's the possibility that the project being saved is that bundle itself, so you remove its resources before copying them to the new _resources_ folder. A temporary folder could be used for the operation, but then there's the issue of guaranteeing an unique name for the temporary folder. So initially I'm submitting this PR with this operation not allowed (and the appropriate feedback to the user explaining it). After it goes through testing and code review and it merged, we can work towards allowing the overwrites if it's a wanted feature (personally I see project bundles as a way to make backups and share projects between different systems, so overwritting doesn't sound like a very important operation on this context).

### For testers

Basically test the overall functionality. Try to save project bundles and test them, check if any resources are missing, try to overwrite existing bundles, saving a project bundle from a project that is also a project bundle. Be careful to work with backup files and in a safe folder.

### For code reviewers

The logic behind this feature is:
- PathUtil now resolves the `local:` base prefix by checking the project file name. If a project isn't open (when running from CLI for example) `local:` is returned again so the method calling the PathUtil method can resolve the path itself.
- `DataFile::writeFile` was rewritten to allow saving with resources depending on the value of a boolean on its parameters. The method that copies the files to the _resources_ folder and updates the paths on the DataFile is `DataFile::copyResources`.
- `main.cpp` checks for the `--makeBundle` command to create a project bundle from the CLI.
- `MainWindow.cpp` and `Song.cpp` were edited so they account for the possibility of saving with resources. For the _Save As_ dialog, a new `saveOption` was added to save as a project bundle.

There are warnings that I left on `DataFile::writeFile` for debugging purposes. Most will be either removed or replaced by an appropriate user warning.

Veratil commented on src/core/DataFile.cpp:
```diff
@@ -326,6 +336,139 @@ bool DataFile::writeFile( const QString& filename )
+	}
+
+	// Creates resources folder
+	if (QDir(resourcesDirName).exists())
+	{
+		qWarning() << "Warning: Resources dir already exists. Proceeding anyways.";
+	}
+	else
+	{
+		if(!QDir().mkdir(resourcesDirName))
+		{
+			qWarning() << "ERROR: Failed to create resources dir!";
+			return false;
+		}
+	}
```
Not sure if this in intention or not, but if the resources directory exists that's okay right? If so, then we can just check for its non-existence.
```suggestion
	if (!QDir(resourcesDirName).exists())
	{
		qWarning() << "resources dir does not exist, creating";
		if (!QDir().mkdir(resourcesDirName))
		{
			qWarning() << "ERROR: Failed to create resources dir!";
			return false;
		}
	}
```

IanCaio replied:
For now I made it Ok, but we can decide whether LMMS should abort creating a bundle if the resources folder already exists. Think it would be safer to abort, what do you think?

Veratil replied:
This would allow only a single bundle to be made at a time, which requires manually moving/removing folders outside of lmms functionality. I don't particularly like that idea. I'd rather we figure out a way to handle it within lmms itself. What if we detect the folder presence and return a list of contents and ask the user what they want to do (abort/clear folder)?

IanCaio replied:
You would still be able to create multiple bundles, just not on the same folder (which kind of makes sense, since the bundle right now is a folder tree basically consisting of the project file and the resources directory). Allowing bundles to be created on the same folder can become a little bit of a mess, since the resources of both projects will get mixed in the same folder. What I think could be improved though is to always create a folder with the project name, inside which there will be the project file and resources dir, which now will be guaranteed to be empty since the folder was just created.

Veratil replied:
What happens if they try to save the same project again?

IanCaio replied:
When I originally thought of this feature I think I had in mind that "Save bundle" would always behave as "Save as". But maybe saving over an existing bundle would be desirable? It still could get messy to allow the same folder to be reused though, we could ask for user confirmation and just clear the folder before copying the new resources. Guess I didn't think as far as that :p

Veratil replied:
> Guess I didn't think as far as that :p

That's what I'm here for. ;)

IanCaio replied:
I was thinking about asking for a bundle name instead of a project file name. The folder would be created and inside it there would be the project file and the resources folder (sort of like how Ardour saves its project files). If the user chooses a name and there's a folder with that name already the operation would be aborted and there would be a warning saying a new name has to be chosen (I thought allowing overwriting could get very dangerous if the user accidentally opts to overwrite the wrong folder, since everything inside it would have to be deleted first). What do you think of that behavior?

Veratil replied:
That should work!

Veratil commented on src/core/DataFile.cpp:
```diff
@@ -326,6 +336,139 @@ bool DataFile::writeFile( const QString& filename )
+	// to avoid name conflicts
+	unsigned int resCounter = 1;
+
+	std::map<QString,std::vector<QString>>::const_iterator it = ELEMENTS_WITH_RESOURCES.begin();
```
Maybe we should make a `typedef` for this `std::map<QString, std::vector<QString>>`.

IanCaio replied:
Oh, forgot to mention, I added a typedef for this. Called it `ResourcesMap`!

Veratil commented on src/core/DataFile.cpp:
```diff
@@ -326,6 +336,139 @@ bool DataFile::writeFile( const QString& filename )
+		++it;
+	}
+
+	// Writes the final DataFile
+	if (fullName.section('.', -1) == "mmpz")
+	{
+		QString xml;
+		QTextStream ts(&xml);
+		write(ts);
+		outfile.write(qCompress(xml.toUtf8()));
+	}
+	else
+	{
+		QTextStream ts(&outfile);
+		write(ts);
+	}
+
+	outfile.close();
+
+	// Make sure the file has been written correctly
+	if (QFileInfo(outfile.fileName()).size() > 0)
+	{
+		// Remove current file if there's one
+		QFile::remove(fullName);
+		// Move temporary file to current file
+		QFile::rename(fullNameTemp, fullName);
+
+		return true;
+	}
+
+	return false;
```
We should consider refactoring this out of the function since it's a near copy/paste of `DataFile::writeFile`.

Maybe the opening code could be in `DataFile::openFile`, this writing part could be `DataFile::writeFile` (with boolean parameter `write_backup`), `DataFile::checkFile` for the written correctly part.

IanCaio replied:
True, I started from the `writeFile` method and ended up reusing too much of it. I'll try to extract parts of it to avoid repetition!

IanCaio commented:
With the latest commits there were a couple of changes to the way project bundles are built:
- Now there's a new folder on the working dir of LMMS, called `projectbundles`
- The `writeBundle` method takes a bundle name, not a file name. A folder with that name will be created inside `projectbundles`, which will contain the bundle itself (project file + resources).
- Bundles currently can't be overwritten, `writeBundle` will fail if there's a folder with that name already (I'm thinking of allowing overwriting later, by recursively deleting the folder of the existing bundle and writing again from scratch).
- The `local:` base prefix uses the `Engine::getSong()->projectFileName()` method to retrieve the local folder. For that reason, when running from CLI (the only way to currently make a bundle) we might run into the issue of not having a project open. To tackle that, resolving `local:` will return `local:` again if there isn't a project open, so the method calling `PathUtil` can handle the situation. `writeBundle` will check if the base prefix is still there after resolving, and if it is it will resolve the local path by itself.
- For `writeBundle` to be able to resolve the local path, a new member variable was added to `DataFile` called `m_fileName`, which will hold the file path if the DataFile was built from a file (or an empty string if not).

IanCaio commented:
I probably have changed the way this PR works too much by now, but I was trying to figure out the best way in terms of simplicity to introduce in the code base. The project bundle folder idea would require lots of extra coding compared to the one used now:

- To save a project bundle, the user should go on the `Save Project As` option. On the dialog window, there are save options (`Discard MIDI connections` and now `Save as Project Bundle (with resources)`). By selecting the `Save as Project Bundle (with resources)` option, the project file will be saved as a project bundle.
- After saving as a bundle, if the user press the `Save` option the project file will be saved normally (not as a bundle). To save again as a bundle the user have to click `Save Project As` once again.
- It's still not possible to overwrite bundles (you can overwrite the project file with a regular save, but you can't overwrite a bundle because the resources folder is already present).

![SaveAs](https://user-images.githubusercontent.com/22241596/97571513-917c6600-19c6-11eb-9e94-2770b326f37b.png)

IanCaio commented:
I think the basic functionality is finished. A little different from the original concept but I had to change things along the way trying to make the feature as simple as possible. Marking as ready for review. Main comment was updated with some more accurate information on the current state of the PR.

IanCaio commented:
As per discussed with @DomClark on the discord channel about the safety of allowing plugins to be bundled (or accessed through local paths), I removed the Vestige plugins from the project bundle, which now only bundle samples.

I also changed `Song::loadProject` so it doesn't load projects that contains local paths to plugins. I'd like others to double check if I covered every place where plugin's path could be on the XML project file. Currently I'm checking the attribute `plugin` on the `<vestige>` element and for each `<effect>` element, I'm checking the child `<key>` elements and `<attribute>` elements inside those. If any contain `name="file"` I check if `value` has a local path.

superpaik commented:
Sorry, I can't test this functionality. Mingw links are broken, and MSVC version crashes when a sample is drag-and-drop to the song-editor.

IanCaio commented:
PR updated to latest `master`, debugging messages removed and small bug on the error logging for the CLI mode is fixed.

JohannesLorenz commented:
I'm still about to review.

For your information, you seem to mosty be compliant to #4922 . I however wonder if we should allow to let LMMS automatically zip the folder (additional option, or even only option). Isn't that more useful for sharing a project?

IanCaio commented:
> I however wonder if we should allow to let LMMS automatically zip the folder (additional option, or even only option). Isn't that more useful for sharing a project?

I think that's possible, and more convenient probably. But if I'm not mistaken we would need to use another library like `zlib` or a wrapper (i.e.: `QuaZIP`) for that. The compression used for `.mmpz` files seems to be limited to byte arrays.

Besides that there would probably be just a little more tweaking to do: Would have to change the bundle save so it doesn't change the project loaded to the new saved file (so it doesn't try to save over a zipped file). Maybe also make a temporary folder to put all the files in before zipping (could be a folder named after a hash, deleted after the zip file was created).

Definitely a good idea, but could we leave it as an enhancement for the feature? I'd have to look on how to use `zlib` before I can dive into trying that and it probably will have to be added as a submodule to the project tree.

Gumichan01 commented:
Hey,

I saw your message in #1257, I took a look on you PR and tested you work.   
I noticed some issues:

- UX issue

When the bundle saving fails, there is no way to check why it failed. In this case, the project I am trying to save uses VST plugins, so I guess it's normal if I can't save it as a bundle. I know that because I read you PR, but a user won't know that.
Maybe you can provide more information about why the app refuses to save the project as a bundle. 

- The bundle format (project file + resources)

This is very painful to save the bundle in a common folder because if a `resources` folder exists, the app won't save the bundle. I understand that you don't want to overwrite existing directories and I agree. This is a good choice.  
However, if you want to save multiples bundles in the same folder (for example because I want to version my bundles) it could be great to automatically create a dedicated directory called <project-name>, put the project file + `resources` in it and compress the directory (or not, if the user doesn't want to).

- File names

I have a very simple project with two samples provided by LMMS named `bass!.ogg` and `kick01.ogg`. When I save the project as a bundle, in the `resources/` directory, I have:

```
1.ogg
2.ogg
```

Ok I get it.

> To avoid name conflicts between files inside the resource folder, all of them are renamed to a number followed by the extension of the file (i.e.: 1.wav, 2.dll).

But how am I supposed to recognize which one is `bass!.ogg` or `kick01.ogg` in this case?  How could I distinguish the same file among 20+ samples? Maybe @ghost's proposal (https://github.com/LMMS/lmms/issues/1257#issuecomment-452089620) can help.

> 
> Duplicate filename scenario:
> 
> Used samples:
> 
> ```
> a/kick.ogg
> a/snare.ogg
> b/kick.ogg
> b/snare.ogg
> ```
> 
> Would become
> 
> ```
> samples/kick.ogg
> samples/snare.ogg
> samples/kick.1.ogg
> samples/snare.1.ogg
> ```

I also noticed that when I load the saved bundle (`lmms <project-name-in-bundle>.mmpz` in the folder containing the project file and `resources`). It says it cannot find `usersample:resources/1.ogg` and `usersample:resources/2.ogg`.

- Empty resources folder

Is it necessary to generate a bundle if it uses no samples? 


I developed an external utility that those something similar to this feature (https://github.com/Gumichan01/lmms-pkg), some elements you integrated are not in this program and vice-versa. Maybe we can collaborate on this PR.

IanCaio commented:
Hey man! Thanks for taking the time to look into it!

> When the bundle saving fails, there is no way to check why it failed. In this case, the project I am trying to save uses VST plugins, so I guess it's normal if I can't save it as a bundle. I know that because I read you PR, but a user won't know that.
> Maybe you can provide more information about why the app refuses to save the project as a bundle.

True, I left some `qWarnings` but should have converted them to user error messages instead. Having VST plugins on the project shouldn't make the bundle save fail though. LMMS fails loading a project that has VSTs with `local:` paths (which shouldn't happen in normal situations since the bundle doesn't save VSTs with local paths. this is to prevent users from opening malicious project bundles), but it should save projects with VSTs normally keeping their current path. Things that could make the saving fail is having a resources folder already (there's a message box for that one), read/write permissions to create the resources folder and read/write permissions to copy the files that the project has (if some track points to a sample that doesn't exist it will fail too).

> However, if you want to save multiples bundles in the same folder (for example because I want to version my bundles) it could be great to automatically create a dedicated directory called , put the project file + `resources` in it and compress the directory (or not, if the user doesn't want to).

Creating a folder exclusively for the bundle is a good idea, makes it more convenient. Compressing the directory would be more workful since it requires using an external library that would probably have to be added as a submodule to LMMS's tree, maybe we can leave the compressing for after the refactor as an enhancement?

> But how am I supposed to recognize which one is `bass!.ogg` or `kick01.ogg` in this case? How could I distinguish the same file among 20+ samples? Maybe @ghost's proposal ([#1257 (comment)](https://github.com/LMMS/lmms/issues/1257#issuecomment-452089620)) can help.

That could be changed, probably creating a list of all the names used already and appending `-1`, etc (I'm not sure all systems allow multiple dots on the names) on all repeating ones.

> I also noticed that when I load the saved bundle (`lmms <project-name-in-bundle>.mmpz` in the folder containing the project file and `resources`). It says it cannot find `usersample:resources/1.ogg` and `usersample:resources/2.ogg`.

That one is odd, did you save the project again with a regular save? By doing a regular save LMMS could have recalculated the paths and messed up by using sample prefixes instead of `local:` ones. Currently bundles can only be saved with `Save as`, so if you edited a bundle and added new samples they would not be in the resources folder if you hit `Save`. But if the regular `Save` is messing up the paths of the samples already converted to locals I have to look into it. For now I should probably disable the regular `Save` for bundle project files, or at least trigger a warning that it won't keep local paths for new samples.

> Is it necessary to generate a bundle if it uses no samples?

True, I didn't think of that because the action of saving as bundle is explicit, but that should be easy to avoid!

> I developed an external utility that those something similar to this feature (https://github.com/Gumichan01/lmms-pkg), some elements you integrated are not in this program and vice-versa. Maybe we can collaborate on this PR.

That would be really cool! We are in an effort to start a refactor happen and this PR (amongst a few others) are bottlenecks for it to happen, so we should keep it as simple as possible for now yet good enough for a merge and leave enhancements for after the refactor. Your help for improving and reviewing it is greatly appreciated :smile:

IanCaio commented:
@Gumichan01 
Just pushed a small change that fixes one of the raised issues. The bundle now creates its own folder (named after the project file name chosen on the dialog). Bundles still can't be overwritten so if it happens that there's a bundle folder with the same name in the selected folder it will still fail.

Some `qWarnings` that were being used to show errors that happened during saving were now converted to message boxes, so it's easier to figure out what happened.

The currently loaded project remains the original one and if the project was flagged as modified before saving a bundle it keeps flagged as modified. I chose that for 2 reasons:
1) Project bundles right now are designed as a backup, not a continuous production project file (since "Save" won't overwrite the bundle with another bundle). So a modified project should still show as modified even after saving a bundle.
2) It prevents the user from accidentally hitting "Save" and overwriting the bundle's project file, which could cause issues like the one you described of paths being recalculated wrongly (I still would like you to describe me exactly what caused it, so we can try to recreate it with the previous build and figure out why the paths were changed, because I don't think it should have happened)

Next issue I'll tackle is probably the sample names one.

IanCaio commented:
Improved the name conflict handling on last commit. Now file names are kept with an appended number if they repeat themselves. I think that covers most of the points, so far:
- User being warned of why a save failed (`qWarnings` replaced with message boxes, can still work on having more descriptive messages).
- Exclusive folder for the bundle.
- Keeping resources file names.

JohannesLorenz commented:
@IanCaio About using zip, I agree that this can be done easily as a separate PR.

I'm going on with the review...

Gumichan01 commented:
> > I also noticed that when I load the saved bundle (`lmms <project-name-in-bundle>.mmpz` in the folder containing the project file and `resources`). It says it cannot find `usersample:resources/1.ogg` and `usersample:resources/2.ogg`.
> 
> That one is odd, did you save the project again with a regular save?

No I didn't. I just saved it as a bundle, closed and reopened LMMS, and loaded the bundle. I retried to reproduce that just before I wrote this comment and it is not reproduced anymore. Maybe I did it wrong the last time, or maybe it was fixed.

> We are in an effort to start a refactor happen and this PR (amongst a few others) are bottlenecks for it to happen, so we should keep it as simple as possible for now yet good enough for a merge and leave enhancements for after the refactor

Ok 👍

JohannesLorenz reviewed (changes requested):
This is what I have so far. I'll do a complete review another time.

IanCaio commented on src/core/DataFile.cpp:
```diff
@@ -328,6 +405,141 @@ bool DataFile::writeFile( const QString& filename )
+		}
+	}
+
+	// Then check effect plugins
```
Except for Lv2 (which I think uses unique ID for loading) the external instruments can only be loaded through Vestige right? I might need some help to check whether I covered everything, but those were the two places where I found paths to external libraries.

Grepping `local:` strings still leads us to the problem of having to check what they are a path for. But you kind of gave me an idea: What if instead of looking for the local paths for plugin's we just check if there are any local paths for anything other than the `DataFile::ELEMENTS_WITH_RESOURCES` list, and if there are any we don't load the project? That's probably even safer because then instead of blocking some local paths we are blocking **every** local path except the allowed ones.

JohannesLorenz replied:
I like your idea, it should be absolutely safe then, and it's consistent with the other uses of `DataFile::ELEMENTS_WITH_RESOURCES`.

Btw 1: It's correct that Lv2 uses unique IDs (or better called URIDs) for referrencing.
Btw 2: Can this function be made const?

IanCaio replied:
Sure! I'll work on making that change then and make the method `const` while I'm at it

IanCaio replied:
New method to check for unsafe `local:` uses pushed, also changed it to `const` on the last commit!

IanCaio commented on src/core/main.cpp:
```diff
@@ -403,6 +403,28 @@ int main( int argc, char * * argv )
+
+			DataFile dataFile(QString::fromLocal8Bit(argv[i]));
+
+			if (argc > i+1) // Project bundle file name given
```
Isn't that true for the `upgrade` action as well? I just copied the way it handles obtaining the input and output files from it IIRR. If we need that level of sanitization maybe we should leave it for a separate issue to fix the other actions as well?

JohannesLorenz replied:
Hmm it would be better to check these things, but you can do it later if you want. Or fix it now on master and then rebase your code and fix it on your branch. I leave it to you.

IanCaio replied:
Could we do it later? Then we can think of how we should sanitize it both for `upgrade` and `makebundle` and fix both at once. We should fix it but it's not that urgent I think, since few users actually use the CLI and those will usually be more experienced to know that that kind of input could go wrong :smile:

JohannesLorenz replied:
OK. resolving this comment.

IanCaio commented on src/core/Song.cpp:
```diff
@@ -983,9 +983,37 @@ void Song::loadProject( const QString & fileName )
+	}
+	else
+	{
+		// We check if VST plugins contain local paths to prevent malicious code being
```
Changed for just "plugins"

IanCaio commented on src/core/PathUtil.cpp:
```diff
@@ -31,6 +34,23 @@ namespace PathUtil
+			}
 			default                   : return QString("");
 		}
 		return QDir::cleanPath(loc) + "/";
```
Not sure, it sometimes complains about methods that are too large, maybe it's that? Though by LMMS's standards that method isn't even that big :laughing:

JohannesLorenz replied:
I think in many cases, this means that there's a recursion, but I don't see how you could have introduced one...

JohannesLorenz replied:
As said in discord, let's ignore this.

IanCaio commented on src/core/main.cpp:
```diff
@@ -403,6 +403,28 @@ int main( int argc, char * * argv )
 
 			return EXIT_SUCCESS;
 		}
+		else if (arg == "makeBundle" || arg == "--makeBundle")
```
Done! Only thing left is adding it to bash-completion: I've to check how it's done so I'll leave it for a separate commit, then someone can confirm I did it right.

IanCaio commented on src/core/Song.cpp:
```diff
@@ -1180,54 +1208,42 @@ bool Song::saveProjectFile( const QString & filename )
-	}
-
-	return saveResult;
+	return guiSaveProjectAs(m_fileName);
```
Got it, the change was very linked to the feature though, not exactly a refactoring. I had to move things around a bit to make it easier to add the `withResources` parameter to the project save. Maybe a separate commit would be better but I think I needed that change to make the feature work so it might have gone in a single commit.

It ended up making the two methods cleaner I think: Before, `saveAs` called `save` changing the project name first to the new project name. Now `save` calls `saveAs` with the same project name, which I think makes more sense.

JohannesLorenz replied:
There are just many questions I have. Before, the two methods were much different, so putting them together probably changed some logic?

IanCaio replied:
It changed slightly, I'll try to go through the changes and write down why they were made and post it here or on Discord! I'll have to read everything to remember.

Was a bit stuck yesterday working on the `hasLocalPlugin` method. I need to use recursion to go through all XML nodes but I think using a lambda function with recursion is more complicated than I thought, so I'll probably rewrite the method itself as recursive :grimacing:

JohannesLorenz replied:
If you write it here or in Discord, future coders who find bugs in refactoring of this function will not find the description. Such justifications should IMO always go into the commit message.

IanCaio replied:
I'll write it in the PR body. Since we squash and merge those commit messages are usually lost anyways, except for the PR log. I tend to be detailed on my commit messages, but I didn't mention those changes specifically probably because I didn't think they were very critical. IIRR the changes were more organizational then functional, just to make calling the modified `DataFile::writeFile` with the new parameter possible while also keeping the bundle exclusive to the `Save as` action.

But I'll still read the code again and write a justification for the changes so it's easily referenced in the PR.

Spekular commented on src/core/PathUtil.cpp:
```diff
@@ -31,6 +34,22 @@ namespace PathUtil
+				{
+					// Keeps the base prefix so the method calling this
+					// can handle the situation.
+					return QString("local:");
```
This feels like a hack. The entire point of this function is to return the location of a base, and yet this returns a prefix. I think this should probably be handled by the caller.

Also, if you did want this here it should return basePrefix(Base::LocalDir) so the prefixes are only defined once.

IanCaio replied:
The caller handles it, but to know that it will have to handle it we need to keep the prefix. An empty string would result in a valid path, so keeping the prefix looked like a good alternative to signal the caller that the path wasn't resolved. This has to be done because at a very particular case (when creating a bundle from the CLI) the `PathUtil` doesn't have access to a `Song` object because it isn't created (we are writing the bundle straight from the `DataFile`).

It's a good idea change it to `basePrefix(Base::LocalDir)` though, because then we don't need to replace it if we change the prefix in the future!

IanCaio replied:
Done! Using `basePrefix` instead now

IanCaio commented:
Explaining some of the changes to `Song.cpp` methods:

- `Song::loadProject()` was changed so it doesn't load projects that potentially have plugins with local paths, which is checked by calling `DataFile::hasLocalPlugins()`.
- `Song::saveProjectFile()` had a parameter added so the project can be saved either normally or as a bundle (with resources).

As for the behavior of `Song::guiSaveProject` and `Song::guiSaveProjectAs` I'll explain the previous behavior and the new one:

#### Previous behavior

- `Song::guiSaveProject` would set the project name to the name with extension, call `saveProjectFile` and if it succeeded set the project modified status to false, then return true if all went well of false if not.
- `Song::guiSaveProjectAs` would first store the current project name to a member variable, set the project name to the name chosen on the "Save as" dialog, call `guiSaveProject`, if it fails restore the `m_oldFileName` and `m_fileName` variables and returns false. If it succeeds, set `m_oldFileName` as the current file and return true.

#### New behavior

That workaround of having to store the current file name on a variable, changing the file name and then calling `guiSaveProject` could be easily removed by having things work the other way around: Instead of `Save As` calling `Save` with a changed file name, `Save` would call `Save As` with the current file name as an argument, and the whole saving logic could be moved to `guiSaveProjectAs`.

- `Song::guiSaveProjectAs` grabs the value of the checkbox on the save dialog to see if the user wants to save the project as a bundle or not, and uses it to call `Song::saveProjectFile`. Whether the save action succeeded or not, the save options are reset to default (that was already done on the previous behavior). That part guarantees that when `guiSaveProject` calls `guiSaveProjectAs` the `withResources` option will be set to false, so the user can only save bundles with `Save As`. The project filename is updated and modified status is set to false **only** if the project save succeeded **and** it was not a bundle. That is because the bundle save is now designed as a backup save, not a continuous production save: Doing a regular save over a bundle project file will not save new resources with local paths, so by keeping the old project name and the project as modified we make sure that the user knows it still have to save the project normally and prevent that it's accidentally saved over the bundle project file.
- `Song::guiSaveProject` simply calls `guiSaveProjectAs` with the current file name.

By moving the saving logic to `guiSaveProjectAs` we could remove the logic of storing the current project name and restoring it if the save failed, make it easier to implement the `withResources` logic and reduce lines of code. `m_oldFileName` was not removed from the class because it's also used on the `loadProject` method for when a load fails.

JohannesLorenz commented on include/DataFile.h:
```diff
@@ -71,7 +72,9 @@ class LMMS_EXPORT DataFile : public QDomDocument
 	void write( QTextStream& strm );
-	bool writeFile( const QString& fn );
+	bool writeFile(const QString& fn, bool withResources = false);
+	bool copyResources(const QString& resourcesDir);
```
Maybe it should be explicitly stated that copyResources can modify the attributes (if files have the same name)? I know that when `Song` loads something, the `DataFile` is discarded anyways, so it's no problem, but for future coders, it might not be obvious that this changes the XML content.

IanCaio replied:
Added a small Doxygen comment that should warn about the DataFile will be modified.

JohannesLorenz commented on include/DataFile.h:
```diff
@@ -71,7 +72,9 @@ class LMMS_EXPORT DataFile : public QDomDocument
 
 	void write( QTextStream& strm );
-	bool writeFile( const QString& fn );
+	bool writeFile(const QString& fn, bool withResources = false);
```
Since writeFile receives the file name anyways, can you just pass it over to `copyResources` instead of introducing `m_filename` for that? Are these two variables always the same?

IanCaio replied:
The file name `writeFile` receives as an argument is the destination file, `m_filename` on the other hand is the file that the `DataFile` object originated from, which is used to resolve the local paths.

JohannesLorenz replied:
I'm not less confused then before your answer :laughing:

IanCaio replied:
`writeFile` writes a `DataFile` object (i.e.: project file XML) to a file (used to save the project), the file it receives as a parameter is the file it will write it to. But a `DataFile` can be constructed from a file as well, which happens when we load a project from a file (`Song::loadProject` and also on `main.cpp` in the `makebundle` or `upgrade` actions for example). That file it originated from is what's stored in `m_filename`, if the `DataFile` didn't originate from a file `m_filename` is an empty string.

`copyResources` might need to resolve the local paths itself if we are running the command from CLI, so it uses `m_filename` to do so. It couldn't use the filename received as an argument to `writeFile` because that would be the file it's saving the bundle **to**, not the file that it's loading the content **from**.

JohannesLorenz replied:
Got it now. Like discussed in PM, `m_filename` is used to *read* the resources before writing copies of them. So we need the origin of the DataFile. Resolving my comment.

JohannesLorenz commented on src/core/PathUtil.cpp:
```diff
@@ -31,6 +34,22 @@ namespace PathUtil
 			case Base::DefaultGIG       : loc = ConfigManager::inst()->userGigDir(); break;
+			case Base::LocalDir:
+			{
+				Song* s = Engine::getSong();
```
```suggestion
				const Song* s = Engine::getSong();
```

JohannesLorenz reviewed (commented):
2nd review. Reviewed everything except the `Song::guiSaveProject` refactor. I'll do that in a final review.

Comments from this review should be all minor.

IanCaio commented on src/core/DataFile.cpp:
```diff
@@ -328,6 +405,164 @@ bool DataFile::writeFile( const QString& filename )
+ * @param QDomElement parent The parent node being interated. When called
+ *        without arguments, this will be an empty element that will be
+ *        ignored (since the second parameter will be true).
+ * @param bool firstCall Defaults to true, and indicates to this recursive
```
Sorry, I forgot to fix the default parameters comment on the CPP file. The default `parent` value is `QDomElement()` (an empty element), not `documentElement()` (because we can't call it without an object, hence can't use it as a default parameter). I used `firstCall` because there's no way to differentiate between the default empty element and an empty element from the XML project file (this was what caused the infinite recursivity I posted on Discord), so I used this little variable to discern between the method being called for the first time and it being called from one of the recursive iterations.

JohannesLorenz replied:
Understood - you recursed infinitely because everytime you hit an empty `QDomElement`, your code did reset the node to the `documentElement`. Resolving.

JohannesLorenz commented on src/core/DataFile.cpp:
```diff
@@ -328,6 +405,164 @@ bool DataFile::writeFile( const QString& filename )
+ *        indicating that we potentially have plugins with local paths that
+ *        would be a security issue. The Song class can then abort loading
+ *        this project.
+ * @param QDomElement parent The parent node being interated. When called
```
Also, is there a reason why you specified the types after `@param` in this comment? I haven't seen that in doxygen yet.

IanCaio replied:
I've seen it before on `AutomationPattern.cpp`, so I thought it was common to have the types in the `@param` comment. I wrote as I've seen them there: "type name description".

JohannesLorenz replied:
If you type `make doc`, you could see how it works. However, it currently does only display the text from the header, which is wrong. Probably, you either need to move it into the header or into the cpp file.

IanCaio replied:
`make doc` didn't work here for some reason, but anyways, you mean that because there's a Doxygen comment on the header and on the cpp file at the same time, only the header comment is being used in the docs? So I should just remove it from the header?

JohannesLorenz replied:
I think that's the reason, yes. Remove it from header or from cpp file. But both at the same time will not work. You should really try to get `make doc` running or report otherwise.

IanCaio replied:
I didn't have `doxygen` installed, my bad :man_facepalming: . I removed the doxygen comment from the header and left only the source file one. It appears to have build into the document properly. It also seems like doxygen indeed identifies the first word after `@param` as the variable type, as it's colored differently.

JohannesLorenz commented:
I just checked the `Song::guiSaveProject...` stuff. It seems to be correct, and I got it with your description now :+1:  When you reworked the other comments, I can probably approve this PR.

IanCaio commented on src/core/DataFile.cpp:
```diff
@@ -328,6 +405,164 @@ bool DataFile::writeFile( const QString& filename )
+ *        method whether this is the first call. If it is it will use the
+ *        root element as the parent.
+ */
+bool DataFile::hasLocalPlugins(QDomElement parent /* = documentElement()*/, bool firstCall /* = true*/) const
```
So far I did some manual tests by editing the project XML by hand and adding `local:` prefixes on attributes that were not in the list. A unit test would probably be nice, but I never wrote one for LMMS so I'd still have to take a look how they are done.

IanCaio commented on src/core/DataFile.cpp:
```diff
@@ -328,6 +405,164 @@ bool DataFile::writeFile( const QString& filename )
+		// still check its children
+		for
+		(
+			ResourcesMap::const_iterator it = ELEMENTS_WITH_RESOURCES.begin();
```
The idea of having a context dependent check is interesting, in case we turn out using those same tags in other contexts for different purposes. As you said it's quite a big change though, would probably have to rethink the `ELEMENT_WITH_RESOURCES` variable. It would have to contain not only the tag and allowed attributes, but also the parent tag. A good idea that we could keep track of to try implementing in the future, for now it's not strictly necessary but it could make the feature even more fail proof in case future changes allow this kind of thing.

I think the allowed tags are not top level though, aren't they children of tags like `<instrument>` and `<sampletrack>`?

JohannesLorenz replied:
Oh correct, I got that wrong, they are not toplevel. OK, let's resolve this.

IanCaio commented on src/core/DataFile.cpp:
```diff
@@ -264,24 +276,89 @@ void DataFile::write( QTextStream & _strm )
-				SongEditor::tr( "Could not open %1 for writing. You probably are not permitted to "
-								"write to this file. Please make sure you have write-access to "
-								"the file and try again." ).arg( fullName ) );
+			showError(SongEditor::tr("Operation denied"),
```
Oh, forgot this comment. I'm not sure, think the difference is a bit subjective in this case. I mean, the program doesn't crash because the project didn't save, it just stays in the project you had before, so I can see a point to using a Warning instead of the Critical we had before. The difference is basically stylistic, maybe the critical message box has a different parent too if I understood right, but that's pretty much it apparently. Is it even worth changing?

JohannesLorenz replied:
No, just wanted to make sure you saw the difference. Feel free to resolve the conversation.

JohannesLorenz commented:
The doc comment is wrong. Doxygen outputs:

```
DataFile.cpp:492: warning: argument 'QDomElement' of command @param is not found in the argument list of DataFile::hasLocalPlugins(QDomElement parent=QDomElement(), bool firstCall=true) const
DataFile.cpp:492: warning: argument 'bool' of command @param is not found in the argument list of DataFile::hasLocalPlugins(QDomElement parent=QDomElement(), bool firstCall=true) const
DataFile.h:77: warning: The following parameters of DataFile::hasLocalPlugins(QDomElement parent=QDomElement(), bool firstCall=true) const are not documented:
  parameter 'parent'
  parameter 'firstCall'
```

The type is only colored to mark it as argument name, which is wrong in this case.

IanCaio commented:
> The doc comment is wrong. Doxygen outputs:
> 
> ```
> DataFile.cpp:492: warning: argument 'QDomElement' of command @param is not found in the argument list of DataFile::hasLocalPlugins(QDomElement parent=QDomElement(), bool firstCall=true) const
> DataFile.cpp:492: warning: argument 'bool' of command @param is not found in the argument list of DataFile::hasLocalPlugins(QDomElement parent=QDomElement(), bool firstCall=true) const
> DataFile.h:77: warning: The following parameters of DataFile::hasLocalPlugins(QDomElement parent=QDomElement(), bool firstCall=true) const are not documented:
>   parameter 'parent'
>   parameter 'firstCall'
> ```
> 
> The type is only colored to mark it as argument name, which is wrong in this case.

Fixed! But we have a big problem, because there are other places in the codebase where the doxygen comments were written that way, which was where I took reference from. This is probably why `make doc` shows a gazillion warnings here, I ended up not even spotting those because of that.

JohannesLorenz commented:
> But we have a big problem

Yeah, I get 473 kB of warnings, this is indeed a "big" problem :smiley: 

The only left comment is writing an optional test. I'm OK with the code. Should we already call out for testers? Or do you think it does not need further testing (there have been tests a while ago)?

IanCaio commented:
@Gumichan01 has tested recently and I tried to manually test after changes like the `hasLocalPlugin` rewrite. Maybe we could leave those automated unit tests for the future as I'd still have to read on how to write them, but could be a good idea to throw it on Discord in case anyone wants to test.

I think it's working fine though, I tried to continuously check it during the commits.

Btw, CircleCI seems to have failed but not sure why, maybe something went wrong in the middle of the tests?

JohannesLorenz commented:
About the CI error:
```
/root/project/src/core/DataFile.cpp: In member function 'bool DataFile::hasLocalPlugins(QDomElement, bool) const':
/root/project/src/core/DataFile.cpp:523:22: error: 'assert' was not declared in this scope
   assert(childElement);
```
I get that one all the time (because my system has `assert` inside the compiler headers somehow, but CI needs `<cassert>`). However, I then always switch to `Q_ASSERT` (`<QtGlobal>`?), to be consistent with other code.

IanCaio commented:
> About the CI error:
> 
> ```
> /root/project/src/core/DataFile.cpp: In member function 'bool DataFile::hasLocalPlugins(QDomElement, bool) const':
> /root/project/src/core/DataFile.cpp:523:22: error: 'assert' was not declared in this scope
>    assert(childElement);
> ```
> 
> I get that one all the time (because my system has `assert` inside the compiler headers somehow, but CI needs `<cassert>`). However, I then always switch to `Q_ASSERT` (`<QtGlobal>`?), to be consistent with other code.

It's good that it triggered an error though, because it uncovered an issue: `childElement` is of type QDomElement, and `assert` was converting it to `bool`. Even if `toElement()` returned a null node, the assertion would probably be true. `Q_ASSERT` on the other hand was more strict about the type and complained about the conversion.

I'm thinking about removing that assert though, as it seems that no other part of the codebase asserts for the conversion to element, and if the node is not an element, `toElement` will return an empty object that can go through the rest of the codeblock with no problems if I understood right. Asserting `childElement.isNull()` would be a problem as well, for the same reason that caused those infinite iterations (some elements might actually be null, as in empty).

IanCaio commented:
Last commit fixes a bug where in certain situations a `local:` path could be assigned to some resource if it turned out to be the shortest relative path to it, even on regular saves. The `PathUtil::toShortestRelative` method now skips local paths to prevent that from happening, as those should only be used by the bundle save in the allowed resources.

JohannesLorenz reviewed (approved):
Code is OK until here :+1:

superpaik commented:
Some comments from a user point of view.
- you have to go to "Save As..." and then check the option "Save as a Project Bundle". Ok. But it's kind of weird thought. Imagine, you have a project name test.mmpz, that's saved. You go to Save as Bundle. In the filename box, you save it as test.mmpz (original name), but you check the bundle option. Even thought the dialogue says that the file exists and it will be overwritten (project file, test.mmpz), it's not, only the bundle is saved. That can cause user confusion.
- If I have a project, say test1.mmpz, you go to save as (the normal one) and save it as test2.mmpz, you keep working on test2. But if you do the same but checking the "Save as bundle" you keep working on test1. I understand why, but it's kind of odd.
- On the other hand, if the bundle exists, it cannot overwritten. Why do not have the same behaviour as a project? A message saying that files will be overwritten (in fact, delete all the folder and write it again).
I think it all will be more clear and easy to do if we have an specific option for "Save as Bundle" at menu level.

Also (and I think that unfortunately it's not possible to solve, but maybe you guys have a solution) is that if you use a VST that uses samples (like a sampler -vitala or poise, for example-) those samples are not included in the bundle, so you have to keep track of those samples and include them on the bundle if you wanna share it.

IanCaio commented:
Thanks for testing @superpaik !

I agree some details might be a bit confusing from the user perspective, maybe clarifying some behavior with documentation would help for now. This is more of an initial implementation but the idea is to continuously improve it and who knows, maybe even transition the regular "Save" to that approach. It's just that there are lots of changing parts, so it's better done incrementally IMO. Sorry about the big reply, just trying to answer some of your observations with some details from the coding side :sweat_smile: 

> * you have to go to "Save As..." and then check the option "Save as a Project Bundle". Ok. But it's kind of weird thought. Imagine, you have a project name test.mmpz, that's saved. You go to Save as Bundle. In the filename box, you save it as test.mmpz (original name), but you check the bundle option. Even thought the dialogue says that the file exists and it will be overwritten (project file, test.mmpz), it's not, only the bundle is saved. That can cause user confusion.

On the "Save" dialog we have the option of choosing a file or a folder IIRR. If I used the folder selection, the user would probably have to create the folder and then choose it to save the bundle inside. I used the filename to make it more convenient (avoiding that creating folder step), but it ended up causing this weird behavior where the dialog still thinks the save will happen over the file when in reality it will create a folder with the same name (minus the extension). I'll take a look to see if I can omit the "Overwrite" dialog when the "Save with resources" option is selected, but from a quick look on Qt's `QFileDialog` documentation it seems there's an option for that called `QFileDialog::DontConfirmOverwrite`. The problem is that the options are usually set before creating the dialog and changing it later with `setOption()` doesn't seem to guarantee that the option will take effect if the dialog is already visible, so I'm not sure I can alternate the dialog to ask confirmation to overwrite or ignore it depending on whether "Save with resources" is selected or not. I could try to just use it (worst case scenario it still asks confirmation), but then some systems might have that overwrite confirmation behavior and some not.

> * If I have a project, say test1.mmpz, you go to save as (the normal one) and save it as test2.mmpz, you keep working on test2. But if you do the same but checking the "Save as bundle" you keep working on test1. I understand why, but it's kind of odd.

Yeah, that was purposefully. For now the bundle save is just meant for backup/sharing (some things would have to change for the user to be able to continuously work and save over the same bundle). So the project still points to the original file and keeps marked as modified to force the user to do a regular save too and not lose its work.

> * On the other hand, if the bundle exists, it cannot overwritten. Why do not have the same behaviour as a project? A message saying that files will be overwritten (in fact, delete all the folder and write it again).

Since the bundle is (at least now) meant to be a backup, I ended up not bothering about making it able to overwrite automatically, that not being very common when saving backups. It's also a bit safer to obligate the user to manually delete a bundle before writing over it, because if the user picks a project name that would match an existing folder without the extension they could unwillingly overwrite a folder that wasn't a bundle to begin with. If we decide to implement the overwriting despite that, there are some things to decide, as for example: Should it erase everything inside the previous bundle before overwriting or should it only overwrite the project file and samples? If there are samples with the same name should those be overwritten too or assigned a name with a increased number? That change could maybe use a separate PR to decide those things and find the best approach :smile: 

>   I think it all will be more clear and easy to do if we have an specific option for "Save as Bundle" at menu level.

That could be done! I'll later check how that change would look like on code, if it turns out to be simple I can include in this same PR!

> Also (and I think that unfortunately it's not possible to solve, but maybe you guys have a solution) is that if you use a VST that uses samples (like a sampler -vitala or poise, for example-) those samples are not included in the bundle, so you have to keep track of those samples and include them on the bundle if you wanna share it.

I've to confirm, but I think except for plugins that ship with LMMS the others might store their information encoded in their XML encoded data, making it harder to access and change the paths, at least with the approach used right now. I wonder if those plugins use relative paths themselves though? That would make it easier to workaround the fact that those samples are not included in the bundle, but would still force you to copy by hand to the bundle folder :confused:

Spekular commented:
It's taken me quite a while to get around to, but I'd like to review this again tomorrow.

Spekular commented:
I fell behind schedule again so I won't actually do as much of a review as I hoped today, but as I suspected we don't need to hack getBaseDir. Ian wrote on Discord:

> When it's called from the GUI, Engine::getSong returns an object so PathUtil can resolve local: paths. If the user runs lmms makebundle project.mmp bundle.mmp though, the bundle is created from the DataFile but it doesn't actually create a Song, so instead DataFile uses the name of the file it originated from to resolve the local path

But `DataFile.cpp` contains tons of `if(gui)` checks already. We could simply replace `if (PathUtil::baseLookup(resPath) == PathUtil::Base::LocalDir)` with `if (!gui)` in `DataFile::copyResources` and state that the LocalDir prefix is only safe to use when a gui exists. (Sidenote: this feels like a situation where my original plan of using a new class for enhanced paths would have been useful. Then the paths could be initialized with the location of "local" (default to empty string, perhaps) and they would always be able to return it. A `withLocal` method would return a copy of the path adapted to a different local dir, allowing it to change.)

Secondly, I don't think we should blindly ignore the local path in `PathUtil::toShortestRelative`. Either we add a method that takes a list of allowed bases (in order of descending priority) and pass a list w/o local when necessary, or we add a boolean toggle. Normal projects should be allowed to use local as well, this was one of my long term goals with the enhanced paths (that I left as a later enhancement).

IanCaio commented:
> But `DataFile.cpp` contains tons of `if(gui)` checks already. We could simply replace `if (PathUtil::baseLookup(resPath) == PathUtil::Base::LocalDir)` with `if (!gui)` in `DataFile::copyResources` and state that the LocalDir prefix is only safe to use when a gui exists.

Having PathUtil return an invalid path (not removing the `local:` prefix) sounds like a better alternative IMO than allowing it to return a valid but wrong path (removing `local:` but not resolving it at all). The latter could lead to hard to track bugs if someone used an unresolved path while the first would be easier to track since it would probably result in an error instead. Why is it bad to return the unresolved path in the situations where it can't be resolved?

> Secondly, I don't think we should blindly ignore the local path in `PathUtil::toShortestRelative`. Either we add a method that takes a list of allowed bases (in order of descending priority) and pass a list w/o local when necessary, or we add a boolean toggle. Normal projects should be allowed to use local as well, this was one of my long term goals with the enhanced paths (that I left as a later enhancement).

Only a few elements are allowed to have local paths and LMMS will not load any project that has local paths outside that list. If we simply allow local paths to be considered on `toShortestRelative`, VSTs could try to use it and we would have a project that can't be opened later because LMMS would refuse to, that's why I removed it from there. I agree we could later come up with a way for the method consider local paths for those allowed elements, but right now it's easier to just not allow it at all and leave it just for the bundle save.

The way local paths works right now is not set in stone, so we can slowly change and improve it in the future to transition our projects to use it too (besides the bundles), but there are things to consider while doing so that I believe are out of the scope of this PR, we'd probably be able to cover them better by doing it incrementally.

Spekular commented:
> Having PathUtil return an invalid path (not removing the `local:` prefix) sounds like a better alternative IMO than allowing it to return a valid but wrong path (removing `local:` but not resolving it at all). The latter could lead to hard to track bugs if someone used an unresolved path while the first would be easier to track since it would probably result in an error instead. Why is it bad to return the unresolved path in the situations where it can't be resolved?

In either scenario the caller has to be aware of the behavior, which is only specified in the actual method implementation. It makes sense to me that if there's no file saved yet, we assume the local directory is as high up in the hierarchy as possible. I find `local:` to be more surprising as a return value than an empty string, and want to minimize surprise. After some more though, the best solution for now is probably to add a status indicator to the method, moving it back towards a more self-documenting state. If the function only takes a base and returns a path, there's no indication that something could go wrong. If it accepts a base and a pointer to a status bool/int, you're alerted to the fact that something could go wrong.

> Only a few elements are allowed to have local paths and LMMS will not load any project that has local paths outside that list. If we simply allow local paths to be considered on `toShortestRelative`, VSTs could try to use it and we would have a project that can't be opened later because LMMS would refuse to, that's why I removed it from there. I agree we could later come up with a way for the method consider local paths for those allowed elements, but right now it's easier to just not allow it at all and leave it just for the bundle save.
> The way local paths works right now is not set in stone, so we can slowly change and improve it in the future to transition our projects to use it too (besides the bundles), but there are things to consider while doing so that I believe are out of the scope of this PR, we'd probably be able to cover them better by doing it incrementally.

Incremental improvement is good and all, but I think we need to be careful to avoid "it gets worse before it gets better" changes. Again, `toShortestRelative` currently does exactly what it says on the tin. Excluding `local:` is surprising, adding an `allowLocal` argument would remove the element of surprise (you can once again figure out the behavior via name and type).

IanCaio commented:
I didn't see it much as "it gets worse before it gets better", but "it's just decent, but it will get better" :laughing: 

I like the ideas you suggested though, would you be good with the changes if I make `getBaseDir` accept an optional status pointer and use it instead, and also add a `allowLocal` parameter to `toShortestRelative` and just default it to `false` so it isn't allowed right now?

Spekular commented:
> [...] would you be good with the changes if I make `getBaseDir` accept an optional status pointer and use it instead, and also add a `allowLocal` parameter to `toShortestRelative` and just default it to `false` so it isn't allowed right now?

Absolutely!

Spekular commented on src/core/PathUtil.cpp:
```diff
@@ -14,8 +14,11 @@ namespace PathUtil
-	QString baseLocation(const Base base)
+	QString baseLocation(const Base base, bool* ok /* = nullptr*/)
 	{
+		// ok is true unless something goes wrong
```
I believe 0 is used for success in most error codes, so it would be better to have a `bool* error`. That way you can have several different error codes (which all evaluate to true) and one success code (which evaluates to false). If you flip it, you can only have one error, or you sacrifice the ability to use `if(error)`.

Of course this isn't an issue in practice right now, since we only have one error code, but I think it'd be more consistent to follow that convention.

IanCaio replied:
I'm flipping it now, should I replace it with a pointer to `int` or can we keep it as a `bool` for now while we don't have different error codes?

Spekular replied:
Either is fine with me, so maybe let's stick with the bool for now. Either way we don't end up in a situation where a bunch of if statements have to be inverted because we want to add more status codes.

IanCaio replied:
Cool, it's done then!

Spekular commented on src/core/PathUtil.cpp:
```diff
@@ -131,28 +136,37 @@ namespace PathUtil
 	}
 
-	QString toShortestRelative(const QString & input)
+	QString toShortestRelative(const QString & input, bool allowLocal /* = false*/)
```
I'm a bit confused, why are the default arguments commented out?

Veratil replied:
If they're defined in the header this will cause an error saying they're already defaulted or something like that.

Spekular replied:
Oh yeah, I totally missed that they're defined there. Thanks!

Spekular reviewed (approved):
Concerns I had are fixed!

IanCaio commented:
Awesome! If nobody's against it I'm merging it tomorrow.

@superpaik Just letting you know, I was working on the UI change to have a separate menu for the bundle save, but it involved changing the PR further and would delay the reviewing a little bit. We'll still take it into consideration and improve it, just in a separate PR so we can get to the reorg sooner!

superpaik commented:
Perfect! Thanks!
